File: python/arete-sdk/arete/chain.py
"""Chain read client (``/chain/*`` HTTP routes).

Port of ``typescript/core/src/chain.ts`` / Rust ``arete_sdk::chain``: the
:class:`ChainClient` protocol with the ten chain read methods and
:class:`HttpChainClient`, the HTTP implementation authenticated through
:mod:`arete.http` with the ``read`` scope.

``u64`` values are decimal strings on the wire (validated as ``^\\d+$`` within
u64) and native ``int`` in Python; account data is base64.
"""

from __future__ import annotations

import base64
import binascii
from dataclasses import dataclass
from typing import Any, Dict, Optional, Protocol
from urllib.parse import quote

from arete.errors import AreteError
from arete.http import AuthTokenTarget, HttpAuthClient, HttpRequestError

U64_MAX = 18_446_744_073_709_551_615

# Server-side cap on one POST /chain/accounts batch.
MAX_CHAIN_BATCH_ADDRESSES = 100

# encodeURIComponent-safe characters (JS parity).
_URI_COMPONENT_SAFE = "-_.!~*'()"


class ChainError(AreteError):
    """A chain read request failed or returned an invalid response."""

    def __init__(
        self,
        message: str,
        *,
        status: Optional[int] = None,
        path: Optional[str] = None,
        body: Any = None,
        code: Optional[str] = None,
    ) -> None:
        super().__init__(message)
        self.status = status
        self.path = path
        self.body = body
        self.code = code


@dataclass(frozen=True)
class ChainClock:
    slot: int
    unix_timestamp: int
    epoch: Optional[int] = None
    leader_schedule_epoch: Optional[int] = None


@dataclass(frozen=True)
class MintAccountInfo:
    address: str
    owner_program: str
    decimals: Optional[int] = None
    supply: Optional[str] = None
    mint_authority: Optional[str] = None
    freeze_authority: Optional[str] = None


@dataclass(frozen=True)
class TokenAccountInfo:
    address: str
    owner_program: str
    mint: Optional[str] = None
    owner: Optional[str] = None
    amount: Optional[str] = None
    ui_amount_string: Optional[str] = None


@dataclass(frozen=True)
class TokenBalanceInfo:
    exists: bool
    owner: str
    mint: str
    amount: str  # raw amount as a decimal string (mirrors the TS surface)
    context_slot: int
    address: Optional[str] = None
    token_program: Optional[str] = None
    decimals: Optional[int] = None
    ui_amount_string: Optional[str] = None


@dataclass(frozen=True)
class NativeBalanceInfo:
    lamports: int
    context_slot: int


@dataclass(frozen=True)
class RawAccountInfo:
    address: str
    owner_program: str
    lamports: int
    executable: bool
    data: bytes


class ChainClient(Protocol):
    """Read access to Solana chain state through the stack's ``/chain/*`` routes."""

    async def exists(self, address: str) -> bool: ...

    async def lamports(self, address: str) -> int: ...

    async def native_balance(
        self, address: str, *, min_context_slot: Optional[int] = None
    ) -> NativeBalanceInfo: ...

    async def minimum_balance_for_rent_exemption(self, space: int) -> int: ...

    async def clock(self) -> ChainClock: ...

    async def account(self, address: str) -> Optional[RawAccountInfo]: ...

    async def accounts(
        self, addresses: list[str]
    ) -> list[RawAccountInfo | None]: ...

    async def mint(self, address: str) -> Optional[MintAccountInfo]: ...

    async def token_account(self, address: str) -> Optional[TokenAccountInfo]: ...

    async def balance(
        self,
        *,
        owner: str,
        mint: str,
        token_program: Optional[str] = None,
        min_context_slot: Optional[int] = None,
    ) -> TokenBalanceInfo: ...


def _encode_component(value: str) -> str:
    return quote(value, safe=_URI_COMPONENT_SAFE)


def _parse_decimal_u64(value: Any, name: str, path: str) -> int:
    if not isinstance(value, str) or not value or not value.isascii() or not value.isdigit():
        raise ChainError(
            f"Invalid chain response for '{path}': {name} must be a decimal u64 string",
            path=path,
        )
    parsed = int(value)
    if parsed > U64_MAX:
        raise ChainError(
            f"Invalid chain response for '{path}': {name} exceeds u64", path=path
        )
    return parsed


def _parse_raw_account(body: Any, path: str) -> RawAccountInfo:
    try:
        data = base64.b64decode(body["data"], validate=True)
    except (binascii.Error, ValueError) as e:
        raise ChainError(
            f"Invalid chain response for '{path}': account data is not valid base64: {e}",
            path=path,
        ) from e
    return RawAccountInfo(
        address=body["address"],
        owner_program=body["ownerProgram"],
        lamports=body["lamports"],
        executable=body["executable"],
        data=data,
    )


def _serialize_context_slot(value: int) -> str:
    if not isinstance(value, int) or isinstance(value, bool) or value < 0 or value > U64_MAX:
        raise ValueError("minContextSlot must be a non-negative integer within u64")
    return str(value)


def _with_context_slot(
    body: Dict[str, Any], min_context_slot: Optional[int]
) -> Dict[str, Any]:
    if min_context_slot is not None:
        body["minContextSlot"] = _serialize_context_slot(min_context_slot)
    return body


class HttpChainClient:
    """HTTP :class:`ChainClient` over a stack base URL.

    ``target`` (a ``solana-gateway-binding`` :class:`AuthTokenTarget`) makes
    every request mint targeted tokens — used by hosted gateway bindings.
    """

    def __init__(
        self,
        base_url: str,
        auth_client: HttpAuthClient,
        *,
        target: Optional[AuthTokenTarget] = None,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._auth = auth_client
        self._target = target

    def _url(self, path: str) -> str:
        return self._base_url + (path if path.startswith("/") else "/" + path)

    async def _request(self, method: str, path: str, body: Any = None) -> Any:
        try:
            return await self._auth.request_json(
                method,
                self._url(path),
                json_body=body,
                target=self._target,
                scopes=("read",),
            )
        except HttpRequestError as e:
            status = getattr(e, "status", None)
            raise ChainError(
                f"Read request to '{path}' failed ({status}): {e}",
                status=status,
                path=path,
                body=getattr(e, "body", None),
                code=getattr(e, "code", None),
            ) from e

    async def exists(self, address: str) -> bool:
        body = await self._request("GET", f"/chain/exists/{_encode_component(address)}")
        return bool(body["exists"])

    async def lamports(self, address: str) -> int:
        body = await self._request("GET", f"/chain/lamports/{_encode_component(address)}")
        return body["lamports"]

    async def native_balance(
        self, address: str, *, min_context_slot: Optional[int] = None
    ) -> NativeBalanceInfo:
        path = "/chain/native-balance"
        payload = _with_context_slot({"address": address}, min_context_slot)
        body = await self._request("POST", path, payload)
        return NativeBalanceInfo(
            lamports=_parse_decimal_u64(body.get("lamports"), "lamports", path),
            context_slot=_parse_decimal_u64(body.get("contextSlot"), "contextSlot", path),
        )

    async def minimum_balance_for_rent_exemption(self, space: int) -> int:
        body = await self._request(
            "GET", f"/chain/rent-exemption/{_encode_component(str(space))}"
        )
        return body["lamports"]

    async def clock(self) -> ChainClock:
        body = await self._request("GET", "/chain/clock")
        return ChainClock(
            slot=body["slot"],
            epoch=body.get("epoch"),
            leader_schedule_epoch=body.get("leaderScheduleEpoch"),
            unix_timestamp=body["unixTimestamp"],
        )

    async def account(self, address: str) -> Optional[RawAccountInfo]:
        path = f"/chain/accounts/{_encode_component(address)}"
        body = await self._request("GET", path)
        return None if body is None else _parse_raw_account(body, path)

    async def accounts(
        self, addresses: list[str]
    ) -> list[RawAccountInfo | None]:
        requested = list(addresses)
        if len(requested) > MAX_CHAIN_BATCH_ADDRESSES:
            raise ValueError(
                f"addresses exceeds the {MAX_CHAIN_BATCH_ADDRESSES}-address "
                "limit for one batch"
            )
        if not requested:
            return []
        path = "/chain/accounts"
        body = await self._request("POST", path, {"addresses": requested})
        # Items are positionally aligned with the requested addresses.
        return [
            None if item is None else _parse_raw_account(item, path)
            for item in body["items"]
        ]

    async def mint(self, address: str) -> Optional[MintAccountInfo]:
        body = await self._request(
            "GET", f"/chain/mints/{_encode_component(address)}"
        )
        if body is None:
            return None
        return MintAccountInfo(
            address=body["address"],
            owner_program=body["ownerProgram"],
            decimals=body.get("decimals"),
            supply=body.get("supply"),
            mint_authority=body.get("mintAuthority"),
            freeze_authority=body.get("freezeAuthority"),
        )

    async def token_account(self, address: str) -> Optional[TokenAccountInfo]:
        body = await self._request(
            "GET", f"/chain/token-accounts/{_encode_component(address)}"
        )
        if body is None:
            return None
        return TokenAccountInfo(
            address=body["address"],
            owner_program=body["ownerProgram"],
            mint=body.get("mint"),
            owner=body.get("owner"),
            amount=body.get("amount"),
            ui_amount_string=body.get("uiAmountString"),
        )

    async def balance(
        self,
        *,
        owner: str,
        mint: str,
        token_program: Optional[str] = None,
        min_context_slot: Optional[int] = None,
    ) -> TokenBalanceInfo:
        path = "/chain/balances"
        payload: Dict[str, Any] = {"owner": owner, "mint": mint}
        if token_program is not None:
            payload["tokenProgram"] = token_program
        payload = _with_context_slot(payload, min_context_slot)
        body = await self._request("POST", path, payload)
        return TokenBalanceInfo(
            exists=body["exists"],
            address=body.get("address"),
            owner=body["owner"],
            mint=body["mint"],
            token_program=body.get("tokenProgram"),
            amount=body["amount"],
            decimals=body.get("decimals"),
            ui_amount_string=body.get("uiAmountString"),
            context_slot=_parse_decimal_u64(body.get("contextSlot"), "contextSlot", path),
        )

File: typescript/core/src/chain.ts
export interface ChainClock {
  slot: number;
  epoch?: number;
  leaderScheduleEpoch?: number;
  unixTimestamp: number;
}

export interface MintAccountInfo {
  address: string;
  ownerProgram: string;
  decimals: number | null;
  supply: string | null;
  mintAuthority: string | null;
  freezeAuthority: string | null;
}

export interface TokenAccountInfo {
  address: string;
  ownerProgram: string;
  mint: string | null;
  owner: string | null;
  amount: string | null;
  uiAmountString: string | null;
}

export interface TokenBalanceInfo {
  exists: boolean;
  address: string | null;
  owner: string;
  mint: string;
  tokenProgram?: string;
  amount: string;
  decimals?: number | null;
  uiAmountString?: string | null;
  contextSlot: bigint;
}

export interface NativeBalanceInfo {
  lamports: bigint;
  contextSlot: bigint;
}

export interface ContextSlotOptions {
  minContextSlot?: number | bigint;
}

export interface TokenBalanceInput {
  owner: string;
  mint: string;
  tokenProgram?: string;
}

export interface RawAccountInfo {
  address: string;
  ownerProgram: string;
  lamports: bigint;
  executable: boolean;
  data: Uint8Array;
}

export interface ChainClient {
  exists(address: string): Promise<boolean>;
  lamports(address: string): Promise<number>;
  nativeBalance(address: string, options?: ContextSlotOptions): Promise<NativeBalanceInfo>;
  minimumBalanceForRentExemption(space: number): Promise<number>;
  clock(): Promise<ChainClock>;
  account(address: string): Promise<RawAccountInfo | null>;
  accounts(addresses: readonly string[]): Promise<(RawAccountInfo | null)[]>;
  mint(address: string): Promise<MintAccountInfo | null>;
  tokenAccount(address: string): Promise<TokenAccountInfo | null>;
  balance(input: TokenBalanceInput, options?: ContextSlotOptions): Promise<TokenBalanceInfo>;
}

type FetchLike = typeof fetch;

function joinUrl(baseUrl: string, path: string): string {
  return `${baseUrl.replace(/\/$/, '')}${path.startsWith('/') ? path : `/${path}`}`;
}

function decodeBase64(encoded: string): Uint8Array {
  if (typeof atob === 'function') {
    const binary = atob(encoded);
    const bytes = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) {
      bytes[i] = binary.charCodeAt(i);
    }
    return bytes;
  }
  const bufferCtor = (globalThis as { Buffer?: { from(input: string, encoding: string): Uint8Array } }).Buffer;
  if (bufferCtor) {
    return new Uint8Array(bufferCtor.from(encoded, 'base64'));
  }
  throw new Error('No base64 decoder available in this environment');
}

const MAX_BATCH_ADDRESSES = 100;

interface RawAccountBody {
  address: string;
  ownerProgram: string;
  /** Decimal string, not a number: a balance above 2^53 would be rounded by `JSON.parse`. */
  lamports: string;
  executable: boolean;
  data: string;
}

function toRawAccount(body: RawAccountBody | null): RawAccountInfo | null {
  if (!body) {
    return null;
  }
  return {
    address: body.address,
    ownerProgram: body.ownerProgram,
    lamports: decimalU64(body.lamports, 'lamports'),
    executable: body.executable,
    data: decodeBase64(body.data),
  };
}

function decimalU64(value: string, field: string): bigint {
  if (!/^\d+$/.test(value)) {
    throw new TypeError(`${field} must be a decimal u64 string`);
  }
  const parsed = BigInt(value);
  if (parsed > 18_446_744_073_709_551_615n) {
    throw new RangeError(`${field} exceeds u64`);
  }
  return parsed;
}

function serializeContextSlot(value: number | bigint): string {
  if (typeof value === 'number') {
    if (!Number.isSafeInteger(value) || value < 0) {
      throw new RangeError('minContextSlot must be a non-negative safe integer or bigint');
    }
    return String(value);
  }
  if (value < 0n || value > 18_446_744_073_709_551_615n) {
    throw new RangeError('minContextSlot must fit in u64');
  }
  return value.toString();
}

function withContextSlot<T extends object>(input: T, options?: ContextSlotOptions): T & {
  minContextSlot?: string;
} {
  if (options?.minContextSlot === undefined) {
    return input;
  }
  return {
    ...input,
    minContextSlot: serializeContextSlot(options.minContextSlot),
  };
}

export function deriveHttpEndpoint(wsUrl: string): string {
  try {
    const parsed = new URL(wsUrl);
    if (parsed.protocol === 'ws:') {
      parsed.protocol = 'http:';
    } else if (parsed.protocol === 'wss:') {
      parsed.protocol = 'https:';
    }
    return parsed.toString().replace(/\/$/, '');
  } catch {
    return wsUrl.replace(/^wss?:/i, (protocol) => protocol.toLowerCase() === 'wss:' ? 'https:' : 'http:');
  }
}

export function createChainClient(httpBaseUrl: string, fetchImpl: FetchLike): ChainClient {
  return {
    async exists(address: string): Promise<boolean> {
      const path = `/chain/exists/${encodeURIComponent(address)}`;
      const response = await fetchImpl(joinUrl(httpBaseUrl, path));
      const body = await parseReadResponse<{ exists: boolean }>(response, path);
      return body.exists;
    },

    async lamports(address: string): Promise<number> {
      const path = `/chain/lamports/${encodeURIComponent(address)}`;
      const response = await fetchImpl(joinUrl(httpBaseUrl, path));
      const body = await parseReadResponse<{ lamports: number }>(response, path);
      return body.lamports;
    },

    async nativeBalance(
      address: string,
      options?: ContextSlotOptions
    ): Promise<NativeBalanceInfo> {
      const path = '/chain/native-balance';
      const response = await fetchImpl(joinUrl(httpBaseUrl, path), {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify(withContextSlot({ address }, options)),
      });
      const body = await parseReadResponse<{ lamports: string; contextSlot: string }>(
        response,
        path
      );
      return {
        lamports: decimalU64(body.lamports, 'lamports'),
        contextSlot: decimalU64(body.contextSlot, 'contextSlot'),
      };
    },

    async minimumBalanceForRentExemption(space: number): Promise<number> {
      const path = `/chain/rent-exemption/${encodeURIComponent(String(space))}`;
      const response = await fetchImpl(
        joinUrl(httpBaseUrl, path)
      );
      const body = await parseReadResponse<{ lamports: number }>(response, path);
      return body.lamports;
    },

    async clock(): Promise<ChainClock> {
      const path = '/chain/clock';
      const response = await fetchImpl(joinUrl(httpBaseUrl, path));
      return parseReadResponse<ChainClock>(response, path);
    },

    async account(address: string): Promise<RawAccountInfo | null> {
      const path = `/chain/accounts/${encodeURIComponent(address)}`;
      const response = await fetchImpl(joinUrl(httpBaseUrl, path));
      return toRawAccount(await parseReadResponse<RawAccountBody | null>(response, path));
    },

    async accounts(addresses: readonly string[]): Promise<(RawAccountInfo | null)[]> {
      // Copied before anything else: `readonly string[]` accepts a mutable array, so a caller can
      // splice it while the request is in flight and the cardinality check below would then compare
      // the response against a list that is no longer what was asked for. The Python client copies
      // for the same reason.
      const requested = [...addresses];
      if (requested.length > MAX_BATCH_ADDRESSES) {
        throw new RangeError(
          `addresses exceeds the ${MAX_BATCH_ADDRESSES}-address limit for one batch`
        );
      }
      if (requested.length === 0) {
        return [];
      }
      const path = '/chain/accounts';
      const response = await fetchImpl(joinUrl(httpBaseUrl, path), {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify({ addresses: requested }),
      });
      const body = await parseReadResponse<{ items: (RawAccountBody | null)[] }>(response, path);
      // A different count would shift every later account onto the wrong address.
      if (body.items.length !== requested.length) {
        throw new TypeError(
          `Invalid chain response for '${path}': expected ${requested.length} items, got ${body.items.length}`
        );
      }
      // Positionally aligned with `requested`; absent accounts arrive as null.
      return body.items.map(toRawAccount);
    },

    async mint(address: string): Promise<MintAccountInfo | null> {
      const path = `/chain/mints/${encodeURIComponent(address)}`;
      const response = await fetchImpl(joinUrl(httpBaseUrl, path));
      return parseReadResponse<MintAccountInfo | null>(response, path);
    },

    async tokenAccount(address: string): Promise<TokenAccountInfo | null> {
      const path = `/chain/token-accounts/${encodeURIComponent(address)}`;
      const response = await fetchImpl(joinUrl(httpBaseUrl, path));
      return parseReadResponse<TokenAccountInfo | null>(response, path);
    },

    async balance(
      input: TokenBalanceInput,
      options?: ContextSlotOptions
    ): Promise<TokenBalanceInfo> {
      const path = '/chain/balances';
      const response = await fetchImpl(joinUrl(httpBaseUrl, path), {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify(withContextSlot(input, options)),
      });
      const body = await parseReadResponse<
        Omit<TokenBalanceInfo, 'contextSlot'> & { contextSlot: string }
      >(response, path);
      return {
        ...body,
        contextSlot: decimalU64(body.contextSlot, 'contextSlot'),
      };
    },
  };
}
import { parseReadResponse } from './read';
